tests/e2e/graceful_shutdown_test.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"go.etcd.io/etcd/tests/v3/framework/interfaces"
+)
+
+func TestGracefulShutdown(t *testing.T) {
+	clusterSize := 3
+	testRunner := e2e.NewE2eRunner()
+	testRunner.BeforeTest(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	clus := testRunner.NewCluster(ctx, t, config.WithClusterSize(clusterSize))
+	// clean up orphaned resources like closing member client.
+	defer clus.Close()
+	// shutdown each etcd member process sequentially
+	// and start from old leader, (new leader), (follower)
+	tryShutdownLeader(ctx, t, clus.Members())
+}
+
+// tryShutdownLeader tries stop etcd member if it is leader.
+// it also asserts stop leader should not take longer than 1.5 seconds and leaderID has been changed within 500ms.
+func tryShutdownLeader(ctx context.Context, t *testing.T, members []interfaces.Member) {
+	quorum := len(members)/2 + 1
+	for len(members) >= quorum {
+		leader, leaderID, term, followers := getLeader(ctx, t, members)
+		stopped := make(chan error, 1)
+		go func() {
+			// each etcd server will wait up to 1 seconds to close all idle connections in peer handler.
+			start := time.Now()
+			leader.Stop()
+			took := time.Since(start)
+			if took > 1500*time.Millisecond {
+				stopped <- fmt.Errorf("leader stop took %v longer than 1.5 seconds", took)
+				return
+			}
+			stopped <- nil
+		}()
+
+		// etcd election timeout could range from 1s to 2s without explicit leadership transfer.
+		// assert leader ID has been changed within 500ms
+		time.Sleep(500 * time.Millisecond)
+		// validate leader ID has been changed
+		resps, err := followers[0].Client().Status(ctx)
+		require.NoError(t, err)
+		require.NotEqualf(t, resps[0].Leader, leaderID, "expect old leaderID %x changed to new leader ID %x", leaderID, resps[0].Leader)
+		require.Equal(t, resps[0].RaftTerm, term+1)
+
+		err = <-stopped
```

Comment:
I'm thinking that we should check the new leaderID after current leader stopped, just in case that flaky caused by limit resource in github action. 

```go

go func() {
     defer close(stopped)
     leader.Stop()
}()

select {
case <-stopped:
case <-time.After(5 * time.Second):
      t.Fatalf("timeout")
}

followers[0].Client().Status()
...
```

Since it is not performance test, we should focus on function-level. Does it make senses here?

Reply:
Leader stop takes at least 1 seconds while leadership expires after `from 1s to 2s` naturally. If we check the new leader ID after current leader stopped, it does not guarantee leader changed due to graceful shutdown.

Reply:
Thanks for the comment. It makes sense to me.